<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Ingest kallisto output — ingest_kallisto • AssembleSC</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Ingest kallisto output — ingest_kallisto" />
<meta property="og:description" content="Shortcircuits the regular ingest() function to return a SingleCellExperiment
object derived from regular kallisto output. This is used typically for processing
plate-based data, where each well corresponds to a single cell, and each cell
is in its own directory, e.g. sample folder." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">AssembleSC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Ingest kallisto output</h1>
    
    <div class="hidden name"><code>ingest_kallisto.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Shortcircuits the regular <code><a href='ingest.html'>ingest()</a></code> function to return a <code>SingleCellExperiment</code>
object derived from regular kallisto output. This is used typically for processing
plate-based data, where each well corresponds to a single cell, and each cell
is in its own directory, e.g. sample folder.</p>
    </div>

    <pre class="usage"><span class='fu'>ingest_kallisto</span>(
  <span class='no'>dir.all</span>,
  <span class='kw'>.id</span> <span class='kw'>=</span> <span class='st'>"auto"</span>,
  <span class='kw'>strategy</span> <span class='kw'>=</span> <span class='st'>"full_length"</span>,
  <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='kw'>NULL</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dir.all</th>
      <td><p>A string with the path pointing to the directory <em>above</em> the individual sample directories,
the one containing <em>all</em> samples (cells) from the run; this is due to between-sample
normalization that needs to be performed prior to analysis. Required.</p></td>
    </tr>
    <tr>
      <th>.id</th>
      <td><p>A character vector specifying the sample ids. Default is 'auto', which deduces the
sample id based on the folder names within the directory specified by <code>dir.all</code>.</p></td>
    </tr>
    <tr>
      <th>strategy</th>
      <td><p>A string specifying the sequencing protocol's strategy, used to instruct
the normalization approach. Default is '3prime_tagged'.</p></td>
    </tr>
    <tr>
      <th>tx2gene</th>
      <td><p>A two-column data.frame linking transcript id (column 1) to gene id (column 2).
Required for gene-level summarisation, when <code>txOut = FALSE</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function utilizes under the hood the <code><a href='https://rdrr.io/pkg/tximport/man/tximport.html'>tximport::tximport()</a></code> function to
read in the counts data, performing scaling, gene summarisation, and the like
as specified. For more details, see <code>?tximport::tximport()</code>.</p>
<p>Note that depending on the sequencing protocol, different counts-normalization
strategies may need to be used. For example, for full-length 5'-&gt;3' sequencing,
gene/transcript length normalization is a common step, as opposed to 3' targeted
strategy which does not require this step.</p>
<p>For 3' tagged RNA-seq, correcting the counts for gene length will induce a bias
in your analysis, because the counts do not have a length bias. Thus, it is
recommended to use the original counts, as a counts matrix.
For full-length sequencing, its recommended to set <code>countsFromAbundance</code> to</p>
<p>For more information, see:
<a href='https://bioconductor.org/packages/devel/bioc/vignettes/tximport/inst/doc/tximport.html#downstream_dge_in_bioconductor'>Downstream DGE in Bioconductor from tximport</a> and <a href='https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6178912/'>Swimming downstream</a>.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other ingest: 
<code><a href='ingest.html'>ingest</a>()</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='no'>dir</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata'</span>, <span class='st'>'aligners'</span>, <span class='st'>'kallisto'</span>, <span class='st'>'Sample_A1'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'AssembleSC'</span>)
<span class='no'>t2g.file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata'</span>, <span class='st'>'metadata'</span>, <span class='st'>'tx2gene_human.tsv'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'AssembleSC'</span>)
<span class='no'>t2g</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>readr</span><span class='kw ns'>::</span><span class='fu'><a href='https://readr.tidyverse.org/reference/read_delim.html'>read_tsv</a></span>(<span class='no'>t2g.file</span>, <span class='kw'>col_names</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='fu'>ingest_kallisto</span>(<span class='no'>dir</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Robert Amezquita.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


