<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Ingest of scRNA-seq Alignment Results — ingest • AssembleSC</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Ingest of scRNA-seq Alignment Results — ingest" />
<meta property="og:description" content="Reads in the resulting count matrices following alignment of scRNA-seq data.
Supported aligners are detailed below, as each has their own way
of organizing results." />
<meta property="og:image" content="/logo.png" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">AssembleSC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Ingest of scRNA-seq Alignment Results</h1>
    
    <div class="hidden name"><code>ingest.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Reads in the resulting count matrices following alignment of scRNA-seq data.
Supported aligners are detailed below, as each has their own way
of organizing results.</p>
    </div>

    <pre class="usage"><span class='fu'>ingest</span>(
  <span class='kw'>dir</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>aligner</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>mode</span> <span class='kw'>=</span> <span class='st'>"standard"</span>,
  <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"raw"</span>,
  <span class='kw'>strategy</span> <span class='kw'>=</span> <span class='st'>"3prime_tagged"</span>,
  <span class='kw'>.id</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>.sep</span> <span class='kw'>=</span> <span class='st'>"."</span>,
  <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='kw'>NULL</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dir</th>
      <td><p>A string with the path to the sample directory where a scRNA-seq experiment results
from an aligner are stored. Required. For the <code>kallisto</code> aligner, must be the directory containing all
cells (samples) from the scRNA-seq run.</p></td>
    </tr>
    <tr>
      <th>aligner</th>
      <td><p>A string specifying a supported aligner from one of the following:
'STARsolo', 'kb-python'. For 'kallisto', see <code>[ingest_kallisto()]</code> for more details. Required.</p></td>
    </tr>
    <tr>
      <th>mode</th>
      <td><p>A string specifying the mode with which the aligner was run.
'standard' refers to kallisto/kb-python default (and is the same as 'Gene' for
<code>aligner = STARsolo</code>), 'lamanno' refers to a kb-python specific mode, and 'Gene',
'GeneFull', 'and 'Velocyto' refer to STARsolo specific modes. Default is
'standard'.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>A string specifying whether to use the raw or aligner-filtered count
matrices. Default is 'raw'.</p></td>
    </tr>
    <tr>
      <th>strategy</th>
      <td><p>A string specifying the sequencing protocol's quantification strategy. Default is '3prime_tagged'.</p></td>
    </tr>
    <tr>
      <th>.id</th>
      <td><p>A string that serves to uniquify samples by prepending to the barcode.
Default is NULL, which means no id is prepended to barcodes. If .id = 'auto',
the .id string is deduced from the basename of the <code>dir</code> argument.</p></td>
    </tr>
    <tr>
      <th>.sep</th>
      <td><p>The separator between <code>.id</code> and the droplet/cell barcode, default is a full stop.</p></td>
    </tr>
    <tr>
      <th>tx2gene</th>
      <td><p>A two column data.frame where the first column is the transcript id, and the
second column is the corresponding gene id. Required for when the aligner is set to kallisto.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An annotated <code>SingleCellExperiment</code> object.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Below is a list of aligners that are supported or with planned support,
and details on how they are run and their respective organization of results:</p><ul>
<li><p>kb-python (kallisto|bustools) (standard):</p><ul>
<li><p>Run via kb-python workflow via <code>kb count</code> with <code>--filter bustools</code> enabled.</p></li>
<li><p>Alignment results are organized as follows:</p><ul>
<li><p><code>{sample}/{counts_filtered | counts_unfiltered}/{cells_x_genes.barcodes.txt, cells_x_genes.genes.txt, cells_x_genes.mtx}</code></p></li>
</ul></li>
<li><p>Alignment statistics are stored in: <code>{sample}/run_info.json</code></p></li>
</ul></li>
<li><p>kb-python (kallisto|bustools) (--lamanno) not supported yet:</p><ul>
<li><p>Run via kb-python workflow via <code>kb count</code> with <code>--lamanno</code> and <code>--filter bustools</code> flags enabled.</p></li>
<li><p>Alignment results are organized as follows:</p><ul>
<li><p><code>{sample}/{counts_filtered | counts_unfiltered}/{cells_x_genes.barcodes.txt, cells_x_genes.genes.txt, cells_x_genes.mtx}</code></p></li>
</ul></li>
<li><p>Alignment statistics are stored in: <code>{sample}/run_info.json</code></p></li>
</ul></li>
<li><p>STARsolo:</p><ul>
<li><p>Run via STAR in STARsolo mode with with the <code>--soloFeatures Gene GeneFull SJ Velocyto</code> and <code>--soloCellFilter</code> enabled.</p></li>
<li><p>Alignment results are organized as follows:</p><ul>
<li><p><code>{sample}/Solo.out/{mode}/{filtered | raw}/{barcodes.tsv, features.tsv, matrix.mtx}</code></p></li>
</ul></li>
<li><p>Alignment statistics are stored in: <code>{sample}/Log.final.out</code></p></li>
<li><p>Note that running STARsolo with Velocyto feature output only provides a raw, unfiltered output. Not yet supported.</p></li>
</ul></li>
<li><p>kallisto:</p><ul>
<li><p>Run via <code>kallisto quant</code>.</p></li>
<li><p>Alignment results are organized as follows:</p><ul>
<li><p><code>{sample}/{abundance.tsv | abundance.h5}</code></p></li>
</ul></li>
<li><p>Alignment statistics are stored in: <code>{sample}/run_info.json</code></p></li>
<li><p>See <code>[ingest_kallisto()]</code> for a custom function designed for kallisto runs, since the input is
quite different compared to STARsolo/kb-python.</p></li>
</ul></li>
</ul>

<p>Another consideration is the sequencing capture strategy. The reason this is important is
for downstream normalization of the counts, specifically whether to adjust counts for transcript-length.
In many cases with scRNA-seq, the strategy is 3' tagged, which does not require any gene/transcript
length correction as the counts do not have a length bias. However, as is often the case for
plate-based protocols such as SMARTseq2 with full-length sequencing, then indeed the counts
should be adjusted. The parameter <code>capture</code> designates whether the strategy was '3prime_tagged' (default),
which does not perform length-correction, versus 'full_length', which does.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>ingest_kallisto</p>
<p>Other ingest: 
<code><a href='ingest_kallisto.html'>ingest_kallisto</a>()</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>dat.path</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata/aligners/STARsolo/Sample_A1/'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'AssembleSC'</span>)

<span class='fu'>ingest</span>(<span class='no'>dat.path</span>,
       <span class='kw'>aligner</span> <span class='kw'>=</span> <span class='st'>'STARsolo'</span>, <span class='kw'>mode</span> <span class='kw'>=</span> <span class='st'>'Gene'</span>,
       <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>'filtered'</span>,
       <span class='kw'>.id</span> <span class='kw'>=</span> <span class='st'>'A1'</span>, <span class='kw'>.sep</span> <span class='kw'>=</span> <span class='st'>'.'</span>)</div><div class='output co'>#&gt; <span class='message'>Transposing matrix to features (rows) x barcodes (cols).</span></div><div class='output co'>#&gt; class: SingleCellExperiment 
#&gt; dim: 100 100 
#&gt; metadata(0):
#&gt; assays(1): counts
#&gt; rownames(100): CUSG00000000001.1 ENSG00000223972.5 ...
#&gt;   ENSG00000127054.20 ENSG00000283712.1
#&gt; rowData names(1): .row_id
#&gt; colnames(100): A1.AAACCTGAGCCTTGAT A1.AAACCTGAGGACATTA ...
#&gt;   A1.AACTCAGTCATCTGCC A1.AACTCCCAGAAGGCCT
#&gt; colData names(2): barcode .id
#&gt; reducedDimNames(0):
#&gt; altExpNames(0):</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Robert Amezquita.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


